dm/syncer/dml.go

```diff
@@ -683,14 +685,34 @@ func genKeyList(table string, columns []*model.ColumnInfo, dataSeq []interface{}
 	return buf.String()
 }
 
+// truncateIndexValues truncate prefix index from data.
+func truncateIndexValues(indexColumns *model.IndexInfo, tiColumns []*model.ColumnInfo, data []interface{}) []interface{} {
+	values := make([]interface{}, 0, len(indexColumns.Columns))
+	for i, iColumn := range indexColumns.Columns {
+		tcolumn := tiColumns[i]
+		if data[i] != nil && (tcolumn.Tp == mysql.TypeVarchar || tcolumn.Tp == mysql.TypeVarString ||
```

Comment:
```suggestion
		if data[i] != nil && tcolumn.EvalType() == types.ETString
```

seems this is enough?

Reply:
`EvalType()` includes several additional cases, like `TypeUnspecified`,`TypeNull`,`TypeNewDate`,`TypeGeometry`. 
I'm afraid it may has  problem.

Reply:
i found `TruncateIndexValue` alreadt do precheck in 
https://github.com/pingcap/tidb/blob/fde18397028a3ab30f38ab4ef964b6ef8e0adf7a/tablecodec/tablecodec.go#L1290

so we can just simply remove those check?

Reply:
yes, `TruncateIndexValue` will check too. i was intended to do this check to avoid to convert all kinds data to `Datum`, because the prefix index does not  use requently. 
as `Datum` can conveniently convert to data `interface{}`, this check is seems redundant just like your comment, so i will remove this check.